src/screenshot.cpp

```diff
@@ -856,3 +863,82 @@ bool MakeScreenshot(ScreenshotType t, const char *name)
 
 	return ret;
 }
+
+static byte _owner_colours[OWNER_END + 1];
+
+/**
+ * Return the colour a tile would be displayed with in the small map in mode "Owner".
+ *
+ * @param tile The tile of which we would like to get the colour.
+ * @return The colour of tile in the small map in mode "Owner"
+ */
+static inline byte GetTopviewPixels(TileIndex tile)
+{
+	Owner o;
+
+	switch (GetTileType(tile)) {
+	case MP_INDUSTRY: o = OWNER_END;          break;
+	case MP_HOUSE:    o = OWNER_TOWN;         break;
+	default:          o = GetTileOwner(tile); break;
+		/* FIXME: For MP_ROAD there are multiple owners.
+		 * GetTileOwner returns the rail owner (level crossing) resp. the owner of ROADTYPE_ROAD (normal road),
+		 * even if there are no ROADTYPE_ROAD bits on the tile.
+		 */
+	}
+
+	return _owner_colours[o];
+}
+
+static void TopviewScreenCallback(void *userdata, void *buf, uint y, uint pitch, uint n)
+{
+	uint8 *ubuf = (uint8 *)buf;
+
+	uint num = (pitch * n);
+	uint row, col;
+	byte val;
+
+	for (uint i = 0; i < num; i++) {
+		row = y + (int)(i / pitch);
+		col = (MapSizeX() - 1) - (i % pitch);
+
+		TileIndex tile = TileXY(col, row);
+
+		if (IsTileType(tile, MP_VOID)) {
+			val = 0x00;
+		}
+		else {
+			val = GetTopviewPixels(tile);
+		}
+
+		*ubuf = (uint8)_cur_palette.palette[val].b;
+		ubuf += sizeof(uint8); *ubuf = (uint8)_cur_palette.palette[val].g;
+		ubuf += sizeof(uint8); *ubuf = (uint8)_cur_palette.palette[val].r;
+		ubuf += sizeof(uint8);
+		ubuf += sizeof(uint8);
+	}
+}
+
+/**
+ * Make a top-viewed screenshot.
+ * @param name the name to give to the screenshot.
+ */
+bool MakeTopviewWorldScreenshot()
+{
+	/* setup owner table */
+	const Company *c;
+
+	/* fill with some special colours */
+	_owner_colours[OWNER_TOWN]  = MKCOLOUR(0xB4);
```

Comment:
magic numbers are sad :(
Where did these come from? can they be put into an enum somewhere and/or used from somewhere else?

Reply:
Sorry, this code is not made by me. I changed them into proper constants.